statistics/histogram.go

```diff
@@ -108,25 +108,34 @@ func (hg *Histogram) GetUpper(idx int) *types.Datum {
 	return &d
 }
 
-// AvgColSize is the average column size of the histogram.
-func (c *Column) AvgColSize(count int64) float64 {
+// AvgColSize is the average column size of the histogram. These sizes are derived from function `encode`
+// and `Datum::ConvertTo`, so we need to update them if those 2 functions are changed.
+func (c *Column) AvgColSize(count int64, isKey bool) float64 {
 	if count == 0 {
 		return 0
 	}
-	switch c.Histogram.Tp.Tp {
-	case mysql.TypeFloat:
-		return 4
-	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong,
-		mysql.TypeDouble, mysql.TypeYear:
+	// Note that, if the handle column is encoded as value, instead of key, i.e,
+	// when the handle column is in a unique index, the real column size may be
+	// smaller than 8 because it is encoded using `EncodeVarint`. Since we don't
+	// know the exact value size now, use 8 as approximation.
+	if c.IsHandle {
 		return 8
-	case mysql.TypeDuration, mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
-		return 16
-	case mysql.TypeNewDecimal:
-		return types.MyDecimalStructSize
-	default:
-		// Keep two decimal place.
-		return math.Round(float64(c.TotColSize)/float64(count)*100) / 100
 	}
+	histCount := c.TotalRowCount()
+	notNullRatio := 1.0
+	if histCount > 0 {
+		notNullRatio = 1.0 - float64(c.NullCount)/histCount
+	}
+	switch c.Histogram.Tp.Tp {
+	case mysql.TypeFloat, mysql.TypeDouble, mysql.TypeDuration, mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+		return 8 * notNullRatio
+	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeYear, mysql.TypeEnum, mysql.TypeBit, mysql.TypeSet:
```

Comment:
Decimal is missed, is it no need to be considered?

Reply:
According to `EncodeDecimal`, decimal is a variant length type in storage, so its size would be computed using `TotColSize` as string types.